src/rstar.jl

```
@@ -0,0 +1,93 @@
+R = rstar(chn)
+
+# Compute 100 samples of the R⋆ statistic using sampling from according to the prediction probabilities.
+# This approach can be slow and results in a less accurate estimation of the R* statistic.
+# See discussion in Section 3.1.3 in the paper.
+Rs = rstar(chn, 100)
+
+# estimate Rstar
+R = mean(Rs)
+
+# visualize distribution
+histogram(Rs)
+```
+
+## References:
+[Lambert & Vehtari] Ben Lambert and Aki Vehtari. "R∗: A robust MCMC convergence diagnostic with uncertainty using gradient-boostined machines." Arxiv 2020.
+"""
+function rstar(x::AbstractMatrix, y::AbstractVector, nchains::Int, iterations::Int; subset = 0.8, niter = 1_000, eta = 0.5, xgboostparams...)
+
+    N = length(y)
+
+    # randomly sub-select training and testing set
+    Ntrain = round(Int, N*subset)
+    Ntest = N - Ntrain
+
+    ids = Random.shuffle(collect(1:N))
+    train_ids = ids[1:Ntrain]
+    test_ids = ids[(Ntrain+1):end]
+
+    @assert length(test_ids) == Ntest
+
+    # use predicted probabilities?
+    mode = iterations > 1 ? "multi:softprob" : "multi:softmax"
+
+    @info "Training classifier"
+    # train classifier using XGBoost
+    classif = XGBoost.xgboost(x[train_ids,:], niter; label = y[train_ids], 
+                      objective = mode, num_class = nchains,
+                      xgboostparams...)
+
+    @info "Computing R* statistics"
+    Rstats = ones(iterations) * Inf
+    for i in 1:iterations
+        # predict labels for "test" data
+        p = XGBoost.predict(classif, x[test_ids,:])
+
+        pred = if length(p) == Ntest*nchains
+            probs = reshape(p, Ntest, nchains)
+            map(s -> rand(Categorical(s / sum(s))), eachrow(probs))
+        else
+            p
+        end
+
+        # compute statistic
+        a = mean(pred .== y[test_ids])
+        Rstats[i] = nchains*a
+    end
+
+    return Rstats
+end
+
+function rstar(chn::Chains, iterations; kwargs...)
+    nchains = length(chains(chn))
+    @assert nchains > 1
```

Again I guess it would be better to throw an actual error or remove this check.